store/src/java/com/zimbra/cs/index/history/LuceneSearchHistoryIndex.java

```diff
@@ -0,0 +1,172 @@
+            dismax.add(prefixMatch);
+        }
+
+
+        /*
+         * We want the match precedence to be:
+         * 1. Exact match
+         * 2. Term match ("apple" matches "apple juice" or "green apple")
+         * 3. Edge match ("apple" matches "applesauce")
+         * 4. Prefix match ("apple" matches "green apples")
+         *
+         * suffix/infix matches are NOT returned ("apple" should not match "cranapple")
+         *
+         * if we're on a word boundary like "apple ", don't suggest prefix matches
+         */
+        exactMatch.setBoost(10);
+        termMatch.setBoost(5);
+        edgeMatch.setBoost(3);
+        if (!wordBoundary) {
+            prefixMatch.setBoost(1);
+        }
+
+        Analyzer analyzer = new SearchHistoryQueryAnalyzer();
+        try {
+            Reader reader = new HalfwidthKanaVoicedMappingFilter(new StringReader(searchString));
+            TokenStream tokenStream = analyzer.tokenStream(LuceneFields.L_SEARCH_TERMS, reader);
+            CharTermAttribute termAttr = tokenStream.addAttribute(CharTermAttribute.class);
+            try {
+                tokenStream.reset();
+                while (tokenStream.incrementToken()) {
+                    String token = termAttr.toString();
+                    termMatch.add(new TermQuery(new Term(LuceneFields.L_SEARCH_TERMS, token)), Occur.MUST);
+                    if (!wordBoundary) {
+                        prefixMatch.add(new PrefixQuery(new Term(LuceneFields.L_SEARCH_TERMS, token)), Occur.MUST);
+                    }
+                }
+                tokenStream.end();
+                tokenStream.close();
+            } catch (IOException e) {
+            }
+            return dismax;
+        } finally {
+            analyzer.close();
+        }
+    }
+
+    @Override
+    public List<Integer> search(String searchString)
+            throws ServiceException {
+        try {
+            ZimbraIndexSearcher searcher = index.openSearcher();
+            Query query = buildQuery(searchString);
+            ZimbraTopDocs docs = searcher.search(query, null, MAX_INDEX_DOCS);
+            List<Integer> entryIds = new ArrayList<Integer>(docs.getTotalHits());
+            for (ZimbraScoreDoc scoreDoc: docs.getScoreDocs()) {
+                Document doc =  searcher.doc(scoreDoc.getDocumentID());
+                entryIds.add(Integer.parseInt(doc.get(LuceneFields.L_SEARCH_ID)));
+            }
+            return entryIds;
+        } catch (IOException e) {
+            ZimbraLog.search.error("unable to search search history for prefix '%s'", searchString, e);
+            return Collections.emptyList();
+        }
+    }
+
+    @Override
+    public void delete(Collection<Integer> ids) throws ServiceException {
+        List<Integer> idList = new ArrayList<Integer>(ids);
+        try (Indexer indexer = index.openIndexer()) {
+            indexer.deleteDocument(idList, LuceneFields.L_SEARCH_ID);
+            indexer.close();
```

Comment:
"try with resources" calls `close` anyway - so this line is redundant - might even be harmful.
See http://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html for a discussion of the advantages of the newish mechanism (for instance, exceptions from close are suppressed)
Note that the `add` method above uses similar logic - and if an exception was thrown before the close in that, the indexer wouldn't get closed.

Reply:
Ah, I rushed this and forgot to remove the original `indexer.close()` line. Will fix.